File: boj/2667_단지번호붙이기.cpp
/*
1. connected componect
2. visit all nodes in graph
3. not too far, big
BFS
*/


#include <iostream>
#include <queue>
#include <vector>

using namespace std;
#define MAX 26

int n, cnt = 0;
string arr[MAX];
bool visited[MAX][MAX] = {0,};
int dx[4] = {1, -1, 0, 0};
int dy[4] = {0, 0, 1, -1};
vector<int> result;
queue<pair<int, int>> q;


void bfs(int x, int y) {
    q.push({x, y});
    visited[x][y] = true;
    cnt++;

    while(!q.empty()){
        int cx = q.front().first;
        int cy = q.front().second;
        q.pop();

        for(int i=0;i<4;i++){
            int nx= cx + dx[i];
            int ny= cy + dy[i];
            if(0<=nx && 0<=ny && nx<n && ny<n && visited[nx][ny]==false && arr[nx][ny]=='1') {
                q.push({nx, ny});
                visited[nx][ny] = true;
                cnt++;
            }
        }
    }
}

int main(void) {
    cin >> n;

    for(int i=0;i<n;i++) {
        cin >> arr[i];
    }

    for(int i=0;i<n;i++) {
        for(int j=0;j<n;j++) {
            if(arr[i][j]=='1' && visited[i][j]==false) {
                cnt = 0;
                bfs(i, j);
                result.push_back(cnt);
            }
        }
    }

    sort(result.begin(), result.end());

    cout << result.size() << "\n";
    for (int i=0;i<result.size();i++) {
        cout << result[i] << "\n";
    }
}

File: boj/2667_단지번호붙이기.py
'''
1. connected component, 각각의 경로마다 특징을 저장
2. DFS
3. count recursive/ global, visted use/unuse
4. BFS
5. Implement the solution
'''
def dfs(x, y):
    global count
    count += 1
    visited[x][y] = True
    for i in range(4):
        nx, ny = x+dx[i], y+dy[i]
        if 0 <= nx < N and 0 <= ny < N and not visited[nx][ny] and graph[nx][ny] == 1:
            dfs(nx, ny)

N = int(input())
graph = [list(map(int, input())) for _ in range(N)]

visited = [[False] * N for _ in range(N)]
dx = [0, 1, 0, -1]
dy = [1, 0, -1, 0]

result = []

for i in range(N):
    for j in range(N):
        if graph[i][j] == 1 and not visited[i][j]:
            count = 0
            dfs(i, j)
            result.append(count)
result.sort()
print(len(result))
print(*result, sep='\n')

'''
1. connected component, 각각의 경로마다 특징을 저장
2. DFS
3. Propose a solution
4. BFS
5. Implement the solution
'''
def dfs(x, y, cnt):
    graph[x][y] = 0
    for i in range(4):
        nx, ny = x+dx[i], y+dy[i]
        if 0 <= nx < N and 0 <= ny < N and graph[nx][ny] == 1:
            cnt = dfs(nx, ny, cnt+1)
    return cnt

N = int(input())
graph = [list(map(int, input())) for _ in range(N)]

dx = [0, 1, 0, -1]
dy = [1, 0, -1, 0]
result = []

for i in range(N):
    for j in range(N):
        if graph[i][j] == 1:
            result.append(dfs(i, j, 1))

result.sort()
print(len(result))
print(*result, sep='\n')